// @syncfusion/ej2-staging/src/common.js
'use strict'

var demoNames = {
    'BlazorSource': 'Blazor source',
    'AspCoreSource':'ASP.NET Core/MVC Source',
    'BlazorServerApp': 'Blazor Server NET Core 3.1',
    'BlazorServerApp_NET6': 'Blazor Server NET6',
    'BlazorWasmApp_NET6': 'Blazor WASM NET6',
    'aspnetcore' : 'EJ2 ASP.NET Core NET Core 3.1',
    'aspnetcore-net6' : 'EJ2 ASP.NET Core NET6',
    'aspnetmvc' : 'EJ2 ASP.NET MVC'
}
exports.demoNames = demoNames;

var publishUrl = {
    'BlazorSource':'https://nexus.syncfusion.com/#browse/browse:blazor-{branch}',
    'BlazorServerApp': 'https://sfblazor.azurewebsites.net/{branch}/demos/',
    'BlazorServerApp_NET6': 'https://sfblazor.azurewebsites.net/{branch}/net6/demos/',
    'BlazorWasmApp_NET6': 'https://sfblazor.azurewebsites.net/{branch}/wasm/net6/demos/',
    'aspnetcore' : 'https://ej2npmci.azurewebsites.net/{branch}/aspnetcore/',
    'aspnetcore-net6' : 'https://ej2npmci.azurewebsites.net/{branch}/aspnetcore-net6/',
    'aspnetmvc' : 'https://ej2npmci.azurewebsites.net/{branch}/aspnetmvc/'
}
exports.publishUrl = publishUrl;

var blazorSourcePlatform=['BlazorSource'];
exports.blazorSourcePlatform = blazorSourcePlatform;

var aspCoreSourcePlatform=["AspCoreSource"];
exports.aspCoreSourcePlatform = aspCoreSourcePlatform;

var blazorPlatform = ['BlazorServerApp', 'BlazorServerApp_NET6', 'BlazorWasmApp_NET6'];
exports.blazorPlatform = blazorPlatform;

var corePlatform = ['aspnetcore', 'aspnetcore-net6'];
exports.corePlatform = corePlatform;

var mvcPlatform = ['aspnetmvc'];
exports.mvcPlatform = mvcPlatform;

// Function to get the IST Date and Time from any location.
function GetISTDateAndTime(time = null) {
    var date = new Date();
    var UTFoffset = date.getTimezoneOffset();
    date.setMinutes(date.getMinutes() + UTFoffset);
    var IST = 5.5 * 60;
    date.setMinutes(date.getMinutes() + IST);
    return time === 'time'  ? date.toLocaleString() : date.toLocaleDateString();
}
exports.GetISTDateAndTime = GetISTDateAndTime;
